### PR TITLE
Merge the parser refactor into `v4`

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This is a large refactor of the parsing model, started by @honnibal. The refactor turns the paring model into a more canonical Thinc model. As a result, GPU training performance is much-improved, since backpropagation can be done for all states at once, rather than per parsing step.

### Types of change

Enhancement.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: .github/azure-steps.yml
Comment (shadeMe) [inline .github/azure-steps.yml:73]: Can we uncomment these lines here, or should we wait until this PR is merged?

File: spacy/tests/test_misc.py
Comment (shadeMe) [inline spacy/tests/test_misc.py:107]: Can this be uncommented?

File: spacy/tests/parser/test_parse.py
Comment (shadeMe) [inline spacy/tests/parser/test_parse.py:466]: Same question as above.

File: spacy/pipeline/_parser_internals/ner.pyx
Comment (shadeMe) [inline spacy/pipeline/_parser_internals/ner.pyx:392]: Can be removed?

File: spacy/pipeline/_parser_internals/ner.pyx
Comment (shadeMe) [inline spacy/pipeline/_parser_internals/ner.pyx:472]: Same as above.

File: spacy/pipeline/_parser_internals/ner.pyx
Comment (danieldk) [reply spacy/pipeline/_parser_internals/ner.pyx]: I think these need to be re-instated. So, I did so :). Thanks!

File: spacy/tests/test_misc.py
Comment (danieldk) [reply spacy/tests/test_misc.py]: `PrecomputableAffine` doesn't exist as a separate layer anymore, so I removed this.

File: spacy/tests/parser/test_add_label.py
Comment (shadeMe) [inline spacy/tests/parser/test_add_label.py:126]: We probably want to use `xfail` than `skip` for this test (with the reason set `"no beam parser yet"`), I think.

File: spacy/tests/parser/test_parse.py
Comment (shadeMe) [inline spacy/tests/parser/test_parse.py:415]: Seems like a statement from a debugging session; can probably be removed.

File: spacy/tests/parser/test_ner.py
Comment (danieldk) [reply spacy/tests/parser/test_ner.py]: Matt changed the threshold during development. Since the seed is fixed and the original threshold works again, I changed it back to `0.00001`.

File: spacy/tests/parser/test_parse.py
Comment (danieldk) [reply spacy/tests/parser/test_parse.py]: Removed.

**Comment by shadeMe** [comment]
I think we can now remove some of the commented out lines that reference the `v2` parser, seeing how we won't be porting it back to `spacy-legacy`?

**Comment by danieldk** [comment]
> I think we can now remove some of the commented out lines that reference the `v2` parser, seeing how we won't be porting it back to `spacy-legacy`?

Removed all the references to the old versions.

**Comment by adrianeboyd** [comment]
When did you make the decision not to have legacy support for `spacy.TransitionParser.v1` and `spacy.TransitionParser.v2`? This is extremely breaking.

We can deal with having the output can be slightly different given that the factory behavior is versioned on the spacy version (so `parser`+v3.3.0 may not produce the same output as `parser`+v3.4.0 with identical configs), but I think we should try extremely hard not to break existing v3 configs in v4.

**Comment by danieldk** [comment]
> When did you make the decision not to have legacy support for `spacy.TransitionParser.v1` and `spacy.TransitionParser.v2`? This is extremely breaking.

First the idea was to bring back the older versions, possibly as part of `spacy-legacy`. IIRC we discussed this with Matt and he thought it'd be best to just remove the old version.

Since there are changes to the layout of the parameters, etc. it'd be hard to support the `v1`/`v2` versions with the new code base. So, we'll have to fork what is now in spaCy 3.

**Comment by adrianeboyd** [comment]
I don't care about the model compatibility, just the config compatibility. So this continues to work:

```
[components.parser]
factory = "parser"
learn_tokens = false
min_action_freq = 30
moves = null
scorer = {"@scorers":"spacy.parser_scorer.v1"}
update_with_oracle_cut_size = 100

[components.parser.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "parser"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.parser.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode:width}
upstream = "tok2vec"
```

**Comment by svlandeg** [comment]
We had a similar situation for the `entity_linker`, where a new implementation of the component wasn't compatible anymore with older config versions. What we decided at the time is to keep the old code around in `spacy.pipeline.legacy` and to have a "hack" in the factory to decide when to use this legacy version: https://github.com/explosion/spaCy/blob/master/spacy/pipeline/entity_linker.py#L102

It's ugly, but does mean the old configs can still be used. I also imagine that the legacy parser would probably entail a lot more than just the one Python file like we have for the `entity_linker` 😬

**Comment by danieldk** [comment]
> It's ugly, but does mean the old configs can still be used. I also imagine that the legacy parser would probably entail a lot more than just the one Python file like we have for the `entity_linker` 😬

I'll try and see how far we can get. I think configuration-wise the largest difference is that in the new parser we always use the upper layer. So I guess that means that if the user specifies `use_upper = false`, we have to emit a warning and use the upper layer anyway.

**Comment by danieldk** [comment]
I have added back `spacy.TransitionParser.v2` and (besides adding it to the tests) tried training a model from a configuration generated using [the quickstart](https://spacy.io/usage/training#quickstart).

`spacy.TransitionParser.v1` lives in `spacy-legacy`, so I think we have to update that version in legacy later.

I am not sure what to do about:

* `spacy.PrecomputableAffine.v1`, since this is not a separate model anymore.
* `spacy.TransitionModel.v1`, the model and arguments changed a fair bit.

Though I think these are rarely used directly in spaCy model configurations?

File: spacy/pipeline/_parser_internals/search.pyx
Comment (svlandeg) [inline spacy/pipeline/_parser_internals/search.pyx:288]: adding a review comment here so we don't forget to address this

**Comment by svlandeg** [comment]
> I am not sure what to do about:
> 
>     * `spacy.PrecomputableAffine.v1`, since this is not a separate model anymore.
> 
>     * `spacy.TransitionModel.v1`, the model and arguments changed a fair bit.
> 

`_precomputable_affine.py` could just be moved in its entirety to `spacy-legacy`, no? It doesn't have any internal dependencies I think?

Similarly, can't we just move the the code for `spacy.TransitionModel.v1` to legacy?

It would be good to start that PR for `spacy-legacy` while wrapping up this PR, then we can also bump legacy's version here.

File: spacy/tests/parser/test_parse.py
Comment (svlandeg) [inline spacy/tests/parser/test_parse.py:455]: Did we make a decision on this - I thought we could just move the appropriate code to legacy and keep v1 as is?

Comment (svlandeg) [reply spacy/tests/parser/test_parse.py:455]: The test in legacy itself is not exactly the same thing though right? Having the test also in spaCy ensures that the architecture is in fact available (from legacy) and behaves similar as the other architectures (i.e. is a drop-in). Or am I missing something?

Comment (danieldk) [reply spacy/tests/parser/test_parse.py:455]: Ah, I see! I will try to add it back to the test.

Comment (svlandeg) [reply spacy/tests/parser/test_parse.py:455]: Thanks! Basically the legacy/compat discussion is the main thing I wanted to understand fully before merging this in :-)

Comment (danieldk) [reply spacy/tests/parser/test_parse.py:455]: I added back the test, but commented it out. What I think we need to do:

- Merge the refactor into v4.
- Do a v4 dev release.
- Create a `spacy-legacy` branch for v4.
- This PR against `spacy-legacy`: https://github.com/explosion/spacy-legacy/pull/36
- Do a `spacy-legacy` release.
- Update `v4` to require this legacy release and uncomment the test.

Tested locally and with the `spacy-legacy` PR mentioned above the test passes.

Comment (svlandeg) [reply spacy/tests/parser/test_parse.py:455]: Fun! 😁 

Thanks for spelling this out Daniël, that sounds like a perfect plan!

Comment (danieldk) [reply spacy/tests/parser/test_parse.py:455]: I'll convert it into a ticket, so that we don't forget to do it.

**Review by svlandeg — APPROVED** [review]
This has been a truly epic team effort! So great to finally be able to merge this in 🥳  Nice job, team!

**Pull request merged**
